enterprise/backend/test/metabase_enterprise/query_field_validation/api_test.clj

```diff
@@ -0,0 +1,166 @@
+                           :model/Field      {field-2 :id} {:active   false
+                                                            :name     "FB"
+                                                            :table_id table}
+                           :model/Field      {field-3 :id} {:active   false
+                                                            :name     "FC"
+                                                            :table_id table}
+                           ;; QFs not to include:
+                           ;; - Field is still active
+                           :model/QueryField {}            {:card_id  card-1
+                                                            :field_id (mt/id :orders :tax)}
+                           ;; - Implicit reference
+                           :model/QueryField {}            {:card_id            card-2
+                                                            :field_id           field-1
+                                                            :explicit_reference false}
+                           ;; QFs to include:
+                           :model/QueryField {qf-1 :id}    {:card_id  card-1
+                                                            :field_id field-1}
+                           :model/QueryField {qf-2 :id}    {:card_id  card-2
+                                                            :field_id field-2}
+                           :model/QueryField {qf-3 :id}    {:card_id  card-3
+                                                            :field_id field-3}]
+    (mt/with-premium-features #{:query-field-validation}
+      (f {:card-1 card-1
+          :card-2 card-2
+          :card-3 card-3
+          :card-4 card-4
+          :qf-1   qf-1
+          :qf-2   qf-2
+          :qf-3   qf-3}))))
+
+(defmacro ^:private with-test-setup
+  "Creates some non-stale QueryFields and anaphorical provides stale QueryField IDs called `qf-{1-3}` and their
+  corresponding Card IDs (`card-{1-3}`). The cards are named A, B, and C. The Fields are called FA, FB, FB and they
+  all point to a Table called T.
+
+  `card-4` is guaranteed not to have problems"
+  [& body]
+  `(do-with-test-setup (fn [{:keys [~'qf-1 ~'qf-2 ~'qf-3 ~'card-1 ~'card-2 ~'card-3 ~'card-4]}]
+                         ~@body)))
+
+(def ^:private url "ee/query-field-validation/invalid-cards")
+
+(defn- get!
+  ([] (get! {}))
+  ([params]
+   (mt/user-http-request :crowberto :get 200 (str url "?" (codec/form-encode params)))))
+
+(defn- approx=
+  [expected actual]
+  (nil? (hawk.approx/=?-diff expected actual)))
+
+(defn- none-found?
+  [blacklist actual]
+  (if (nil? blacklist)
+    true
+    (not-any? #(some (partial approx= %) actual) blacklist)))
+
+(defn- resp=
+  "Is the response close enough to what we expect? Accounts for extra data from the app DB that could sneak in. If
+  `unexpected` is provided, ensure that they're *not* present.
+
+  Due to pagination this is slightly fragile, sorry."
```

Comment:
Yeah, ouch!

Reply:
To be a bit less fragile I guess you could walk backwards and forwards though pages where appropriate until you've hit the end or hit records outside the range of what you expect?

Reply:
Fair enough 😢